src/Raven.Server/Documents/DatabasesLandlord.cs

```diff
@@ -565,8 +490,22 @@ public Task<DocumentDatabase> TryGetOrCreateResourceStore(StringSegment database
             }
             finally
             {
-                if (_disposing.IsReadLockHeld)
-                    _disposing.ExitReadLock();
+                release?.Dispose();
+            }
+        }
+
+        private IDisposable EnterReadLockImmediately(StringSegment databaseName)
+        {
+            using (var cts = new CancellationTokenSource())
+            {
+                var awaiter = _disposing.ReaderLockAsync(cts.Token).GetAwaiter();
+                if (awaiter.IsCompleted == false)
+                {
+                    cts.Cancel();
```

Comment:
I don't like this.
what happened if between the `IsCompleted` check and the `Cancel()`, you got the lock? Then you have a hanging lock here.

Reply:
done